### PR TITLE
Stats: add to the Yosemite layer to support new stats

### LINKED ISSUE #1066: Stats: add to the Yosemite layer to support new stats
Ref: p99K0U-1iF-p2 and p99K0U-1ex-p2

To support the new stats, we'll need to add Yosemite classes:
- ReadOnly models
- Storage models
- Store class(es)
- new Action(s)

Task dependencies: #1063, #1064, #1065 will need completed first.

### PR DESCRIPTION
**Pull request by ctarda**
Fixes #1066 

Add support to Yosemite for the stats V4 API.

## Changes 
- Add new Action and Store
- Map Networking model objects to Yosemite model objects
- Add readonly and update extensions
- A couple of minor fixes to Storage (oops 😊 )
- Unit tests

## Testing
- Run the tests, make sure everything is ✅ 
- Make sure the code is 👍

### CODE REVIEW

File: Yosemite/Yosemite/Stores/StatsStoreV4.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        let storageOrderStats = storage.loadOrderStatsV4(siteID: String(readOnlyStats.siteID),
+                                                         granularity: readOnlyStats.granularity.rawValue) ?? storage.insertNewObject(ofType: Storage.OrderStatsV4.self)
+        storageOrderStats.update(with: readOnlyStats)
+        handleOrderStatsItems(readOnlyStats, storageOrderStats, storage)
+        storage.saveIfNeeded()
+    }
+
+    /// Updates the provided StorageOrderStats items using the provided read-only OrderStats items
+    ///
+    private func handleOrderStatsItems(_ readOnlyStats: Networking.OrderStatsV4, _ storageStats: Storage.OrderStatsV4, _ storage: StorageType) {
```
Comment (jaclync) [inline Yosemite/Yosemite/Stores/StatsStoreV4.swift]: nit: `func handleOrderStatsIntervals` might sound more consistent with the schema

File: Yosemite/YosemiteTests/Stores/StatsStoreV4Tests.swift [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        }
+
+        statsStore.onAction(action)
+        wait(for: [expectation], timeout: Constants.expectationTimeout)
+    }
+
+
+    // MARK: - Misc
+
+    func date(with dateString: String) -> Date {
```
Comment (jaclync) [inline Yosemite/YosemiteTests/Stores/StatsStoreV4Tests.swift]: nit: could this be private?

File: Storage/Storage/Tools/StorageType+Extensions.swift [hunk 3]
```diff
@@ -96,6 +96,13 @@ public extension StorageType {
         return firstObject(ofType: OrderStatsV4.self, matching: predicate)
     }
 
+    /// Retrieves the Stored OrderStatsV4interval.
+    ///
+    func loadOrderStatsInterval(interval: String) -> OrderStatsV4Interval? {
+        let predicate = NSPredicate(format: "interval ==[c] %@", interval)
```
[the remark this answers is not in the document]
Comment (ctarda) [reply Storage/Storage/Tools/StorageType+Extensions.swift]: That's my intention 😊

File: Yosemite/Yosemite/Stores/StatsStoreV4.swift [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    override public func registerSupportedActions(in dispatcher: Dispatcher) {
+        dispatcher.register(processor: self, for: StatsAction.self)
+    }
+
+    /// Receives and executes Actions.
+    ///
+    override public func onAction(_ action: Action) {
+        guard let action = action as? StatsActionV4 else {
+            assertionFailure("OrderStatsStoreV4 received an unsupported action")
+            return
```
[the remark this answers is not in the document]
Comment (ctarda) [reply Yosemite/Yosemite/Stores/StatsStoreV4.swift]: It seems like we do. If I recall correctly, an assertionFailure will only stop execution on debug mode, so it seems like a good way to alert us if we are messing the types here (as Action is basically a tag interface).

In fact, in this case, it helped me realise I wrote, on my first pass:
`guard let action = action as? StatsAction else`
(notice how it's the wrong type, and the compiler can not catch this)

File: Yosemite/Yosemite/Stores/StatsStoreV4.swift (hunk 1 above)
Comment (ctarda) [reply Yosemite/Yosemite/Stores/StatsStoreV4.swift]: Oh, copy and paste! 😊

File: Storage/Storage/Model/OrderStatsV4+CoreDataProperties.swift [hunk 5]
```diff
@@ -11,40 +11,23 @@ extension OrderStatsV4 {
     @NSManaged public var siteID: Int
     @NSManaged public var granularity: String
     @NSManaged public var totals: OrderStatsV4Totals?
-    @NSManaged public var intervals: NSOrderedSet?
+    @NSManaged public var intervals: Set<OrderStatsV4Interval>?
```
[the remark this answers is not in the document]
Comment (ctarda) [reply Storage/Storage/Model/OrderStatsV4+CoreDataProperties.swift]: 🤦‍♂

**Comment by ctarda** [comment]
Thanks folks. I think I addressed everything, will merge when the build passes

**Pull request merged**